ENH: remove unneeded spaces in float/bool reprs, fixes 0d str

ahaldane opened the pull request:
See discussion in #8983 

This removes the leading space in the string representation of arrays of floating-point type if it is not needed (ie, all positive values), eg:

old behavior:
```python
>>> np.array([1., 2., 3.])
array([ 1.,  2.,  3.])
```
New behavior:
```python
>>> np.array([1., 2., 3.])
array([1., 2., 3.])
```

This noticebly improves the appearance of 0d arrays, so that `np.array(1.0)` prints as itself, instead of as `np.array( 1.0)` as happens after #8983.

It does the same for boolean arrays with only one element, so 0d boolean arrays print as `array(True, dtype=bool)` instead of `array( True, dtype=bool)`.

(longfloats still have a space I think, but I would rather not tackle that here because I think it requires a big rewrite of longfloat formatter)

ahaldane committed:
TST: update tests to remove unneeded whitespace

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -624,6 +618,7 @@ def fillFormat(self, data):
         from . import numeric as _nc
 
         with _nc.errstate(all='ignore'):
+            hassign = self.sign or any(data < 0)
```
Should be `_nc.any`

eric-wieser replied:
And probably `has_sign` not `hassign`

charris commented:
This could break doc tests. Doc tests are evil, but some use them anyway...

ahaldane commented:
@mhvk just for your info as we decide how to make the repr changes, I ran this PR though astropy tests, and it messes up a *lot* of tests:

```
79 failed, 11325 passed, 574 skipped, 68 xfailed, 1 xpassed, 1 pytest-warnings
```

Most of them are things like 
```
Expected:
    <Quantity 15.0 m / s>
Got:
    <Quantity 15. m / s>
```
or
```
Expected:
    array([ 0. ,  0.4,  0.6,  0.8,  1. ])
Got:
    array([0. , 0.4, 0.6, 0.8, 1. ])
```

mhvk commented:
@ahaldane - yes, but I've already got a PR that fixes those tests: https://github.com/astropy/astropy/pull/6090; and am happy to fix things further if we go that route, as I think this PR made printing better in general.  Just don't want to have to do it for each numpy version!

mhvk commented:
@ahaldane - I clearly did not pay much attention to the PR in which you commented... Even up to redoing your work here in #9144 (now closed). In any case, my comment above stands, if slightly differently: I gladly fix the astropy doctests if we end up with more consistent numpy printing -- after all, I think that after this it is not going to change again... But ideally we do make all changes in one go for 1.14!

mhvk commented:
If we try not to break too many things, one option for this would be to expose the `sign` option in `set_printoptions`, maybe having as options ' ', '+', '' (i.e., match what is used by the format mini-language). Though it would be a bit annoying that there is no way not to break compatibility if this is applied to both integers and float, since for integer the default would have to be '', while for floats it would be ' '.

ahaldane commented:
Updated. 

Also, while we are annoying everyone by changing the spaces in array printing, we should try to do it as thoroughly as possible now to avoid any future changes. Specifically, I'd like to add another commit or two to 1. remove spaces from `longfloat`, and 2. I think the repr of ` np.array([100., 2., -1.])` can still be improved (see unit test for this): We should better account for whether the sign is on the longest value or not, currently this PR still has an unneeded space in this case.

(I also still need to update the release notes and other little things)

eric-wieser commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -234,16 +234,57 @@ def test_formatter_reset(self):
+            assert_equal(str(np.array('café', np.unicode_)), 'café')
+            assert_equal(repr(np.array('café', np.unicode_)),
+                         "array('café',\n      dtype='<U4')")
+            assert_equal(str(np.array('café', np.unicode_)), 'café')
```
This case is identical to the one two above it, isn't it?

ahaldane replied:
hmf not sure how that happened.. will fix

mhvk commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -234,16 +234,57 @@ def test_formatter_reset(self):
-                     "array('2005-02-25', dtype='datetime64[D]')")
+            "array('2005-02-25', dtype='datetime64[D]')")
 
+        # 0d arrays are *not* affected by printoptions
```
This comment is now wrong!

ahaldane commented:
Updated. 

I updated the treatment of signed values so the extra whitespace for the sign only appears if the longest elements have a sign. Release note and tests also updated.

I haven't updated the longfloat spacing, but I think I want to leave that for another possible PR. I'd like to get this one through so upstream projects can get their tests to stop complaining.

mhvk commented on doc/release/1.14.0-notes.rst:
```diff
@@ -44,13 +44,33 @@ Improvements
 ``np.set_string_function``, unlike most other numpy scalars. This is no longer
 the case.
+
+removed an unneeded whitespace in float and bool array printing
```
Idem, and also delete "an"

mhvk commented on numpy/core/arrayprint.py:
```diff
@@ -433,11 +427,11 @@ def array2string(a, max_line_width=None, precision=None,
-        Has no effect, do not use.
-
-        .. deprecated:: 1.14.0
+        outputA
```
Piece missing here. Change to "output correctly."?

mhvk commented on numpy/core/arrayprint.py:
```diff
@@ -650,18 +646,22 @@ def fillFormat(self, data):
             if len(non_zero):
-                precision = max([_digits(x, self.precision, format)
-                                 for x in non_zero])
+                precision = max([_trailing_nonzero(x, self.precision)
```
No need to construct a list inside `max` -- btw, unrelated to this PR, but this seems a really slow way for large arrays...

ahaldane replied:
yeah I worried about that too, but then I worked out that `data` is going to be a small array.

`data` only contains the elements that are actually printed on the screen (ie, not the interior which is replaced by `...`)

mhvk replied:
Ah, that makes sense.

mhvk commented on doc/release/1.14.0-notes.rst:
```diff
@@ -44,13 +44,33 @@ Improvements
-----------------------------------------------------
-0d arrays now use the array2string formatters to print their elements, like
-other arrays. The `style` argument of array2string is now non-functional.
+printing of 0d arrays changed to be more consistent with scalars/ndarrays
```
Trivial, but capitalize first word.

mhvk commented on doc/release/1.14.0-notes.rst:
```diff
@@ -44,13 +44,33 @@ Improvements
+a numpy scalar and returns a string, and ``style(a[()])`` is returned. The
+default value is now ``style=None``.
 
 integer scalars are now unaffected by ``np.set_string_function``
```
Idem.

mhvk reviewed (commented):
Looks good, modulo a few text issues in the changelog, and an unneeded list construction (which was there already, but might as well get rid of it).

ahaldane commented on numpy/core/arrayprint.py:
```diff
@@ -650,18 +645,22 @@ def fillFormat(self, data):
-                                 for x in non_zero])
+                precision = max([_trailing_nonzero(x, self.precision)
+                                 for x in abs_non_zero])
+                precision = min(self.precision, precision)
```
Also, I think this line is unnecessary, will remove.

ahaldane commented:
All right, reworked the `trailing_nonzero` part, hopefully the code is not too dense.

mhvk commented:
Looks good to me.

mhvk commented:
I'd be quite happy to merge this, as it seems a follow-up of gh-8983, and finishes making the print options much more logical. Obviously, this will break more doctests, but it does seem best to at least do it in one go. @eric-wieser, @charris: any objections to merging this?

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -328,28 +328,31 @@ gentype_nonzero_number(PyObject *m1)
     }
-    return ret;
+
+    item_str = PyObject_Str(item);
```
As I think I said before, this isn't safe, since `item` can return a copy of `self`, causing infinite recursion.

What actually falls back to this case anyway?
Seems to me that it would be better to just inherit `object.__repr__`, making if obvious that custom dtype users have work to do.

ahaldane replied:
Right, you are talking about [this](https://github.com/numpy/numpy/pull/9143#discussion_r118132580) comment.

Currently, besides potential user-defined types, only numpy only void types fall back to this. But they no longer will after #8981.

I just tried you suggestion by simply removing `gentype_repr` and `gentype_str`, in the last commit pushed here.

Locally (not in the pushed commit) I also temporarily modified `test_rational.c` to remove the str and repr of rational types as a check, and can confirm I get reasonable behavior:
```python
>>> from numpy.core.test_rational import rational
>>> a = np.array([1,2,3], dtype=rational)
>>> a
array([<rational object at 0x7f37f1b4e918>,
       <rational object at 0x7f37f1b4e918>,
       <rational object at 0x7f37f1b4e918>], dtype=rational)
```

I'm not sure how to implement a test of this without recreating a new `test_rational` type missing the str and repr, which seems a bit heavy-handed.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -641,7 +635,8 @@ def fillFormat(self, data):
             self.large_exponent = 0 < min_val < 1e-99 or max_val >= 1e100
-            self.max_str_len = 8 + self.precision
+            pad_sign = self.sign or _nc.any(non_zero < 0)
+            self.max_str_len = 6 + self.precision + pad_sign
```
Why did this decrease by one?

ahaldane replied:
The simple answer is that there appeared to be an unnecessary space.

The more complex answer is that actually I'm not sure why that extra space was there... I'll take another look.

ahaldane replied:
After investigation, I still don't understand why the extra space was there. 

I count "6" extra characters more than the floating part: `x.e+xx` where `x` are digits.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -521,6 +509,8 @@ def array2string(a, max_line_width=None, precision=None,
         formatter = _formatter
 
+    if style is not None and a.shape == () and a.dtype.names is None:
+        return style(a[()])
```
I'd argue this should be all the way upstream in `ndarray.__str__` - ie, complete ignore `np.setprintfunction` for the 0d string case.

eric-wieser replied:
But note that when it goes there, you can't use `PyArray_ToScalar`, as that would circumvent array subclasses that override `[()]` (like maskedarray)

ahaldane replied:
For discussion, that would be counter to current numpy behavior for all scalar types. That is, currently all 0d arrays are affected by `np.setprintfunction` including string types.

I agree that's somewhat counter to the idea of this PR. In this PR, we are making the `str` of 0d arrays behave more like scalars, and the `repr` more like ndarrays. You are suggesting we make the `str` even more like scalars, by not being affected by `np.setprintfunction`.

In some sense I agree, but on the other hand I am hesitant to change even more behavior, and I think the 0d behavior (even in this PR) is a little strange anyway.

I'm going to be away next week and was hoping to get this PR in before then. Do you mind leaving this point for consideration in another PR?

mhvk replied:
Personally, I'd like to keep the option to treat all arrays -- 0d or not -- the same way, as is now provided by the `style=None` option.

eric-wieser replied:
> Do you mind leaving this point for consideration in another PR?

I've folded those into #9201. There's code in the wild that wants to treat `np.array('Some string')` as if it were `'Some string'` - so really, we shouldn't allow `setstringfunction` to change that

mhvk replied:
OK, so this doesn't affect any `style=None` option in an explicit `array2string`, correct? Do agree that `str(array_scalar_string)` should give a fixed result, which is just the string/unicode/bytes object.

eric-wieser replied:
Is that aimed at me or @ahaldane?

mhvk replied:
I thought both; I was just very confused why we were worried about this line.

eric-wieser replied:
My concern is that that line doesn't fully solve the problem, not that it's worse than what we had before. My other concern is that that change isn't really in line with the title of the PR, but that's not too big a deal if they end up in separate commits after the squash

mhvk commented:
Looks good to me. Do need to raise a new issue to also do `longfloat` before the 1.14 release, otherwise it is just annoyingly inconsistent. Also, as I was confused by the discussion: does the removal of `gentype_repr` require #8981 to be merged right after?

ahaldane commented:
Squashed.

My understanding is that the removal of `gentype_repr` here is fine on its own. #8981 is an additional change to make the `str` of 0d arrays be unchageable by `np.setprintfunction`

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -621,6 +594,31 @@ static PyObject *
+        Py_DECREF(arr);
+    }
+    return ret;
+}
```
Can you bundle the removal of `gentype_str` and `gentype_repr`, and the addition of `voidtype_str` into a single commit (and preferably, PR)? Or does that change by itself cause breakage?

ahaldane replied:
I just did this, these changes are in the first commit.

The first commit alone appears to pass tests, so it looks like I can make it its own PR if desired.

ahaldane committed:
ENH: remove unneeded spaces in float/bool reprs

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -621,6 +594,31 @@ static PyObject *
+        item_str = PyObject_Str(item);
+        Py_DECREF(item);
+        return item_str;
+    }
```
This returns a different result to `1.12` - previously, it gave `(1., 2.)`, now it gives `(1.0, 2.0)`.

Perhaps `void.item()` should not convert fields to python scalars?

ahaldane replied:
It does? I don't see that:
```python
>>> np.zeros(2, dtype='f,f')
array([(0., 0.), (0., 0.)],
      dtype=[('f0', '<f4'), ('f1', '<f4')])
```

Also, this code is inside the non-fields clause so should only affect raw void items.

mhvk replied:
I think we're stuck with `item` -- the whole point of it is to just return python things.

eric-wieser replied:
> Also, this code is inside the non-fields clause so should only affect raw void items.

Missed that, thanks

amueller commented:
@charris why are doctests evil?

rkern commented:
Because they routinely get broken by functionally meaningless formatting changes like this one. :-)

I have more to say on the subject, but this isn't really the forum for a back-and-forth. Happy to discuss it on `scipy-dev` or wherever else, though.

eric-wieser commented:
I'd argue that the purposes of doctests is to tests the docs, not the code - if the format changes, then the docs should be updated to match the new output, rather than using an outdated formatter. Not having doctests risks incorrect examples ending up in documentation.

rgommers commented:
> I'd argue that the purposes of doctests is to tests the docs, not the code

Indeed.

> if the format changes, then the docs should be updated to match the new output, rather than using an outdated formatter.

Not a good idea, lots of work for no gain. Who cares about a space more or less, that doesn't make the doc example incorrect. ``doctest`` is just a bad idea (using it by default just forces you to put ``+SKIPTEST`` crap everywhere).

Scipy has https://github.com/scipy/scipy/blob/master/tools/refguide_check.py#L17 to check that the docs are valid Python and run, which is much better (docstrings break, they don't just change outputs in meaningful ways). Would be good for ``numpy`` (and other projects) to take over that approach.

mhvk commented:
I'm with @eric-wieser on this one - it is useful for the user to see the output and it being consistent, and the numpy documentation is, sadly, not as good as it could be in that regard. That said, white space differences can and should be ignored by default, and so should small differences in floats (astropy has `+FLOAT_CMP` for that purpose). 

p.s. Of course, if there *were* doctests in travis, whoever suggested a formatting change would at least have to change the numpy docs, which would be a proper disincentive to change the format needlessly (the present PR *not* being an example of that!).

ahaldane commented:
Out of curiosity, to detect some of the docstrings/docs which whose output has the extra space we are removing here,  I just ran `grep -InRi '\[ 0\.,' numpy` on the source. 

That grep alone returns 140 lines which, in principle, should all be corrected. And I'm sure there are hundreds more where the first value isn't exactly `0.`.

I don't want to fix them here, but it's good to know what we're getting into in this PR.

ahaldane commented:
New slightly more accurate estimate of doc examples needing a fix, giving 213. Includes any floating point numbers, not just `0.`. My last try for now.

```
$ grep -RiIn '\(matrix\|array\)(\[ [0-9]\+\.[0-9]*,' \
> numpy doc benchmarks branding dist doc  tools README.md | grep -v '-' | wc
    213    1226   14565
```

eric-wieser commented:
Can we split this PR into the 0d fixes, and the spacing stuff? One seems important to get in, while the other is gonna create work

mhvk commented:
@eric-wieser - *both* are going to create work, so to minimize the number of times we cause breakage I  would argue for doing things in one go (or at least in a single release). Indeed, I've been holding off on https://github.com/astropy/astropy/pull/6090 because I didn't want to have to make another such PR and knew discussion was ongoing.

ahaldane commented:
Split off the 0d part in #9332, for your viewing pleasure whether or not we merge them separately.

I'm also going to write a mailing post to make more people aware of this proposal.

rgommers commented:
>  That said, white space differences can and should be ignored by default, and so should small differences in floats (astropy has +FLOAT_CMP for that purpose).

Agred with the should parts. What do you mean by "can be ignored"? If there's a simple ``doctest`` setting that does this (I'm not aware of one), then why not tell people to turn it on and this whole discussion is unnecessary.

Regarding ``+FLOAT_CMP``, that's exactly the kind of thing that should be in ``doctest`` itself by default, in order not to have to litter your docstrings with junk. Why not try to get that done? Or fork ``doctest`` into something more sane?

ahaldane committed:
ENH: "pad_sign" option, also threshold and edgeitems

mhvk commented:
@rgommers - yes, fair enough about moving it upstream -- I made this a new issue (https://github.com/astropy/astropy/issues/6312), since it is not something I feel quite competent at doing.

On the white-space, doctest has `NORMALIZE_WHITESPACE` - which sadly gets one only part of the way, since it treats any sequence of white space characters as equal, but not *no* whitespace. See https://docs.python.org/2/library/doctest.html#doctest.NORMALIZE_WHITESPACE

I guess I agree with the bottom line, that ideally one has doctests with an improved runner.

ahaldane commented:
Rebased on top of the commits in #9332.

I also reworked a lot of the user-customizable parts, and allowed the new non-sign-padding to be turned off. Summary of the last commit:

 * Simplified a lot of code duplication in the option parsing
 * A `sign` argument has been added to array2string (this option was already implemented in the old code, just not user-accessible as an argument)
 * a `pad_sign` option has been added, which controls the new behavior in this PR.

When `pad_sign` is True, it gives the old behavior, except for size-1 arrays where it still omits the sign position. (Maybe I should limit it even more, to 0d arrays?)

When `pad_sign` is False (currently default here), it removes the sign padding everywhere if possible.

This should make it easy for downstream projects to disable a lot of the new behavior for their doctests.

ngoldbaum commented:
Sorry to jump in from the numpy mailing list. Can you mention the new `set_printoptions` keyword in the release notes?

ahaldane commented:
Oh right, definitely.

ahaldane commented:
Added release note, docstrings, and a test.

mhvk commented:
@ahaldane - almost on my way to holidays, but one comment on `pad_sign` - I'd make this a string with options `' '`, `'+'`, and `'-'` - i.e., just like the normal formatting mini-language - include space, plus, or nothing for positive numbers.

ahaldane committed:
ENH: Simplify some code in arrayprint.py

mhvk commented:
It would be nice to reach a decision on these stylistic changes - right now, with astropy we're in a state where for multiple months numpy-dev has been failing because of the previous style changes (#8983), and I've been holding off merging a PR to fix this until it has become clear whether things get reverted or carried a bit further (in this PR). Just to be clear: I much rather finish it rather than revert to the very illogical behaviour we had -- but we should aim to do all doctest breakage in one release.

seberg commented:
I agree, but won't think about it myself now. But added the milestone since that seems important then?

ahaldane commented:
We got a bit of pushback on the list, however I am still in favor of these changes. 

One thought is that we should merge sooner rather than later so we have time to try it out, and revert before release of 1.14 if we change our mind. I think a good number of people (in particular, people with doctests) test numpy master automatically. If we merge now and they have problems we can easily revert, since this is a fairly standalone change. On the other hand a reversion would be pretty annoying for everyone too. (I'm also not sure if a policy of "merge uncertain improvements just to try them out" is a good idea).

One reason we might delay is to try to implement a new doctest runner which ignores spaces. I've wanted to try that but time has been limited...

eric-wieser commented:
Let's also not forget that the pr that started all this introduced some recursive loops in some strange cases, which need to be fixed before a release

charris commented:
@eric-wieser Is there an issue for that?

mhvk commented:
My main other argument for merging is that we already have introduced changes to the formatting - we should really have all those in one release...

eric-wieser commented:
@charris: No, but referenced here: https://github.com/numpy/numpy/pull/9346#discussion_r125166031

ahaldane commented:
@mhvk in the case we don't merge this PR, then we should definitely revert the other formatting changes before release, even if it reintroduces pre-existing minor bugs that were fixed.

ahaldane commented:
@mhvk, it looks like Astropy already implements a doctest runner which tries to ignore small differences in array printing.

I noticed this because I trying to implement a doctest runner than ignores whitespace/least-significant-digits around numbers myself (I have an initial implementation working...). I went to try it out in Astropy to see if it fixed all the failed doctests... and noticed astropy [already has one](https://github.com/astropy/astropy/blob/5e1bb1d8b248e2239b0ece4615caa00d7e21cfde/astropy/tests/output_checker.py#L29)!

Instead of modifying all the astropy tests that fail, would it make more sense to modify the OutputChecker to ignore whitespace around numbers?

ahaldane commented:
Actually, `FLOAT_CMP` option in the astropy `OutputChecker` already does a lot of what we want.  You just need to use it in more places, since it is only used as an optional doctest runner on some of the doctests.

mhvk commented:
Yes, the `FLOAT_CMP` is pretty nice (and we really should upstream it). The main problem with the array representation is that we use the `IGNORE_SPACES` (or something like it) by default, but this treats 1 or more spaces as equal to 1, while what happens here is that often what used to be 1 space becomes 0. Still, point taken that it is probably better to just make the doctest laxer than to edit all our tests!

charris commented:
Needs rebase. I'd do it myself, the old `0d` comment seems outdated and in need of removal, but I want someone more knowledgeable about this aspect to take care of it.

charris commented:
I inted to put this in just to see who screams ;) @ahaldane Could you make a list of other print format changing PRs so we can try to get them in all together, and also making it easier to revert them all if needed.

ahaldane commented:
You mean the 0d comment in the release notes? I think it is correct. This PR modifies the 0d behavior relative to current master.

ahaldane commented:
I'll take care of the rebase and PR summary.

ahaldane commented:
Rebased. 

I see what you meant about the 0d stuff in the release notes, I removed it. I think this PR sets 0d behavior back close to master, so no notes about them here. (They will change in #9332).

ahaldane commented:
Wait, I see there is still a duplicate commit here with #9332. Let me fix it..

ahaldane commented:
OK, done. Rebase is now cleaned up.

ahaldane commented:
Summary of 1.14 PRs related to array printing changes, in order of first proposal:

These first two PRs affect the printing of void scalars, and were the motivation for the later PRs:

 * #8157 (not merged) fixes a segmetation fault related to void-scalars, but causes void scalar str/repr to look pretty weird.
 * #8981 (not merged) makes the str/repr of void scalars look like a bytestring, but now 0d arrays are weird, and the 0d printing code made it very difficult to implement this PR due to a recursive import problems, as well as recursion in the printing code itself.

This motivated trying to fix the 0d array printing code. The solution was

 * #8983 (merged) which makes 0d arrays print using for arrayprint.py formatters. Unfortunately, this causes two new problems: 1. 0d float and bool arrays now have an extra whitespace character, 2. 0d string arrays do not print as a raw string anymore, so do not pass the `s = str(array(s))` test.

Relatedly, we also did:

 * #8963 (merged) which adds recursion prevention in the printing code
 * #8989 (merged) somewhat unrelated to the rest, which fixes printing of object arrays containing lists

To fix the problems #8983 introduced, we then have

 * #9139 (this PR, not merged) which fixes the whitespace problems in 0d arrays by overhauling whitespace in all numpy array reprs.
 * #9332 (not merged) which more carefully implements 0d printing, so str prints like scalars and repr like nd elements.
 * (#9143 - do NOT merge - an alternate but inferior solution that we would close once this PR is merged)

Also, just today I started working on a doctest runner PR to ignore whitespace, #9674.

eric-wieser commented:
Don't forget #9345, which is an outstanding issue introduced by #8983

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -51,24 +51,35 @@
+    'nanstr': 'nan',
+    'infstr': 'inf',
+    'sign': '-',
+    'formatter': None }
```
This overlaps with the changes in #9605. Which should we prefer?

eric-wieser replied:
I think I prefer the use of `dict(edgeitems=)` in that implementation, but `_make_options_dict` is nice in this one

srinivasreddy replied:
NP @eric-wieser.  I am closing the PR.  😄

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -51,24 +51,35 @@
-_summaryEdgeItems = 3     # repr N leading and trailing items of each dimension
-_summaryThreshold = 1000  # total items > triggers array summarization
+    if suppress is not None:
+        options['suppress'] = not not suppress
```
Can this just be `bool(suppress)` now?

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -607,14 +600,23 @@ def _formatArray(a, format_function, rank, max_line_len,
+        if sign is False:
+            sign = '-'
+        if sign is True:
+            sign = '+'
```
Not sure about using `is` on booleans. How about

```
if isinstance(sign, bool):
    sign = '+' if sign else '-'
```

mhvk replied:
On the sign, see https://github.com/numpy/numpy/pull/9139#issuecomment-312456857 - I think we should follow precedent and let it be one of `" ", "+", "-"`

eric-wieser replied:
We already do - but we need to support booleans here for backwrads compatibility

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -349,28 +338,26 @@ def _get_format_function(data, precision, suppress_small, formatter):
     else:
         summary_insert = ""
-        data = ravel(asarray(a))
+        data = ravel(asarray(a)) if a.shape != () else asarray(a)
```
This is kind of weird, and probably deserves a comment. Perhaps the special case should be size==1, not ndim==0, which eliminates the need for this line

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -509,24 +509,17 @@ def array2string(a, max_line_width=None, precision=None,
-        formatter = _formatter
+    overrides = _make_options_dict(precision, threshold, edgeitems,
+                                   max_line_width, suppress_small, None, None,
+                                   sign, formatter)
```
Passing these by keyword might be clearer

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -607,14 +600,23 @@ def _formatArray(a, format_function, rank, max_line_len,
+            sign = '+'
+
+        # for backcompatibility, 0d arrays are not padded
+        if sign == ' ' and data.shape == ():
```
Why do we do `data.shape == ()` here, but `data.size == 0` in `BoolFormat`. Let's just use the latter in both places

mhvk replied:
Ah, there is the answer to my question at the top - I guess the real question is whether we want this back-compatibility. I'd argue no, but only weakly. Maybe an alternative is to have a special `sign` for backward compatibility? E.g., have this be the meaning of `sign=False`?

ahaldane replied:
The motivation was this: If anybody is really unhappy with our whitespace changes, they can get back to virtually the same behavior as in 1.13 by doing `np.set_printoptions(sign=' ')` in their test suite. The 0d special case reproduces 1.13 behavior.

This way they can easily shut up their doctests until they have time to properly fix them.

ahaldane replied:
I suppose we could also add an extra sign option `sign='legacy'` or similar.

eric-wieser replied:
@ahaldane: you're responding to mhvk and not me here, right? What are your thoughts on using `data.size == 0` here instead?

mhvk replied:
I like the `sign='legacy'` option - that way, space can be completely consistent with the standard python format specification.

ahaldane replied:
@mhvk, OK, will add the `sign='legacy'` option. I am a little busy this week so there may be a short delay.

@eric-wieser I wanted to use `data.shape == ()` here to catch only 0-d arrays, and `data.size > 0` in `BoolFormat` to catch both 0d and nd-arrays-of-size-1, eg `np.array([4.])`. The goal was to preserve 0d array behavior in general, but also remove the space in size-1 Bool arrays. Note that 0d arrays have `data.size = 1`

eric-wieser replied:
Why do we want `array([True])` yet `array([ 1.0])` to be the defaults?

It seems like we should be consistent here, and choose either:

*  `size == 1`: `array([True])` and `array([1.0])
* `shape == ()`: `array([ True])` and `array([ 1.0])`

I'd favor the first one, because it lets us keep the model that the formatter gets given the flattened array, rather than _"it's 1D except when it isn't"_

ahaldane replied:
In my mind, the space in bool arays is not a "sign" character, it is a space for alignment. Bools don't have a sign.  

This is different from the space in the float array, which represents the sign position. I think we need to keep in in size-1 arrays, because the space should only appear now if the user specified `sign=' '` (which is not the default), meaning that they explicitly said they wanted a space in the sign position. 

Also, `array([1.0])` *is* the default, since `sign='-'` is the new default in this PR.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -51,24 +51,35 @@
+                       sign=None, formatter=None):
+    """ make a dictionary out of the non-None arguments, plus sanity checks """
+
+    options = {k: v for k,v in locals().items() if v is not None}
```
nit: whitespace

eric-wieser reviewed (commented):
Couple minor points, but this all seems pretty good to me.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -51,24 +51,35 @@
+    'edgeitems': 3,  # repr N leading and trailing items of each dimension
+    'threshold': 1000,  # total items > triggers array summarization
+    'precision': 8,
+    'suppress': False,
```
These names (`_float_output_suppress_small` and `_float_precision`) used to be more descriptive - can you add comments that these only effect floats, and that suppress is something about small values?

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -408,7 +395,8 @@ def wrapper(self, *args, **kwargs):
                  suppress_small=None, separator=' ', prefix="",
-                 style=np._NoValue, formatter=None):
+                 style=np._NoValue, formatter=None, threshold=None,
+                 edgeitems=None, sign=None):
```
Can we add these extra arguments via **kwargs**, rather than supporting them positionally? It makes things a little easier if we expose yet more arguments

eric-wieser replied:
Might be nice to deprecate passing these positionally too

mhvk commented on numpy/core/arrayprint.py:
```diff
@@ -641,53 +647,58 @@ def fillFormat(self, data):
+                inflen = len(_format_options['infstr']) + neginf
+                max_str_len = max(max_str_len, nanlen, inflen)
+
+            signchar = '' if self.sign == '-' else self.sign
```
I think we can just use `self.sign` directly, no?

mhvk commented on numpy/core/arrayprint.py:
```diff
@@ -641,53 +647,58 @@ def fillFormat(self, data):
+                max_str_len = max(max_str_len, nanlen, inflen)
+
+            signchar = '' if self.sign == '-' else self.sign
+            format = '%#' + signchar + '%d.%df' % (max_str_len, precision)
```
From https://docs.python.org/3/library/string.html#format-specification-mini-language, it seems the sign character should precede `#`.

ahaldane replied:
Not sure if this a python bug, but using '-' in the python format specification causes the number to be left-justified.
```python
>>> '% #8.1f' % 1.0
'     1.0'
>>> '%+#8.1f' % 1.0
'    +1.0'
>>> '%-#8.1f' % 1.0
'1.0     '
```

ahaldane replied:
oops that refers to your previsous comment on signchar

ahaldane replied:
It's not a bug, it's what C does too.

mhvk replied:
Very odd:
```
In [4]: '%-#8.1f' % 1.
Out[4]: '1.0     '

In [5]: '{:-#8.1f}'.format(1.)
Out[5]: '     1.0'
```
I guess there is a difference between old and new-style formats (I only use the latter, should have checked before commenting).

ahaldane replied:
Also, the docs you linked to are for the `.format()` mini-language, not for the printf-style formatting.

Printf-style formatting is documented at https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting ,  which indeed says that '-' left fustifies, and also does not require any order to the conversion flags.

mhvk replied:
Here I was thinking the two are the same... Learn something new every day.

ahaldane replied:
Actually, while working on this PR at some point I tried to replace the printf-style formatting by the `.format()` function. I don't remember exactly what was the problem anymore, but I remember I determined that it is actually impossible to format things the way we want using `format()`, and `printf` formatting was more powerful in this case.

mhvk commented on numpy/core/arrayprint.py:
```diff
@@ -641,53 +647,58 @@ def fillFormat(self, data):
-                    return self.special_fmt % ('+' + _nan_str,)
+                nan_str = _format_options['nanstr']
+                if self.sign == '+':
+                    return self.special_fmt % ('+' + nan_str,)
```
Not your fault, but yikes, since when is `nan` known to be positive? Can we omit this?

ahaldane replied:
Hmm removing this actuallt breaks tests, because complex types require a sign in the imag part: `0.+nanj`

mhvk replied:
I guess to avoid that one would need yet another special-case for "really, I always want a sign", which is used for complex. But that seems overly cumbersome for a case where there is no confusion, so let's just let it be.

mhvk commented on numpy/core/arrayprint.py:
```diff
@@ -743,20 +755,22 @@ def __init__(self, precision, sign=False):
-                return '+' + _nan_str
+            nan_str = _format_options['nanstr']
+            if self.sign == '+':
+                return '+' + nan_str
```
As above.

mhvk commented on numpy/core/arrayprint.py:
```diff
@@ -743,20 +755,22 @@ def __init__(self, precision, sign=False):
+                return '+' + nan_str
             else:
-                return ' ' + _nan_str
+                return ' ' + nan_str
```
Again not your change, but this is not the same as for regular float (where no space is added)

ahaldane replied:
Yes we discussed this. I haven't updated LongFloat because it needs a total rewrite. (see the 3rd column in the big set of changed tests below).

mhvk commented on numpy/core/arrayprint.py:
```diff
@@ -743,20 +755,22 @@ def __init__(self, precision, sign=False):
+                    return '+' + inf_str
                 else:
-                    return ' ' + _inf_str
+                    return ' ' + inf_str
```
Should we not omit the space for `sign=='-'`? Solution might be the same as the one suggested for float above. Or was this part of the "let's not touch long yet..."?

mhvk commented on numpy/core/arrayprint.py:
```diff
@@ -641,53 +647,58 @@ def fillFormat(self, data):
+                        return self.special_fmt % ('+' + inf_str,)
                     else:
-                        return self.special_fmt % (_inf_str,)
+                        return self.special_fmt % (inf_str,)
```
Should we not include a space for `sign==' '`? I.e., maybe
```
return self.special_fmt % (('' if self.sign == '-' else self.sign) + inf_str,)
```

ahaldane replied:
I think we are OK because of the way `special_fmt` is defined. If `sign=' '` then the `max_str_len` will be one longer, so we will get the necessary prefix space.

mhvk commented on numpy/core/arrayprint.py:
```diff
@@ -743,20 +755,22 @@ def __init__(self, precision, sign=False):
+                return '-' + inf_str
         elif x >= 0:
-            if self.sign:
+            if self.sign == '+':
```
I'd do all three signs here...

mhvk reviewed (commented):
Mostly small comments, with some risking perhaps too many further iterations.

ahaldane commented:
Besides all the little fixes, I tried converting all the long lists of positional/keyword args everywhere to `**kwds`, and deprecate the positional args.

I'm not 100% sure I like it better. In particular I had to write a big helper function `_parse_array2string_args` to parse the keywords. This was needed because `array2string` has some arguments (`separator`, `prefix`, `style`) which `set_printoptions` does not, and also uses a different name for some arguments (eg, `max_field_width` for `linewidth`). That's annoying. Probably ideally `array2string` and `set_printoptions` would have more similar signatures.

Also, since the positional args were perfectly functional, I'm a little hesitant to deprecate them just because they are ugly. It might annoy users without much real gain.

ahaldane commented:
Oh, the other big change is I added the `sign='legacy'` option.

rgommers commented:
> Also, since the positional args were perfectly functional, I'm a little hesitant to deprecate them just because they are ugly. 

Changing parameters into keywords in ``array2string`` looks like an improvement. The change to ``*args, **kwargs`` in ``set_printoptions`` looks undesirable. The keywords currently don't look particularly ugly to me, but even if they are ugly, that's not a reason to make the signature less discoverable and introduce an unnecessary deprecation warning.

ahaldane commented:
Hmm. I assume you are saying that based on the fact that `array2string` currently (1.13) already has an opaque signature of `(self, *args, **kwargs)`, so we aren't losing anything. 

However, upon examination, I think that that opaque signature is a regression caused by #8963, and furthermore the signature is incorrect because there is no `self` argument. In 1.12 the signature was correct, and #8963 was put in 1.13. `array2string.__doc__` is wrong too.

ahaldane commented:
Ah, my last comment only applies to python2. In python3.3 the `__signature__` attribute was added, and now the signature, doc, source etc of wrapped functions is properly updated. 

Nevertheless, this PR currently makes both the `array2string` and `set_printoptions` signatures opaque in all python versions, while they were both previously  inspectable in py3.3+.

I can fix the signatures in python 3.3+ by updating the `__signature__` attribute. Unfortunately for python2 users, `set_printoptions` signature will become opaque, like `array2string` currently is. Any other opinions on whether we should avoid that?

ahaldane commented:
You know what, I would like to avoid reorganizing all the keyword arguments in this PR, it is big enough as it is, with lots of tricky deprecation issues. I'd like to consider those changes for a future PR.

I've just removed those commits, and added a commit that fixes the lost array2string signature. We might also separate this last commit into another PR.

You can see/download the old keyword-reorganizing code in my branch (here)[https://github.com/numpy/numpy/compare/master...ahaldane:whitespace_fix_reorg].

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -349,66 +341,38 @@ def _get_format_function(data, precision, suppress_small, formatter):
+        if a.shape == () and options.get('sign', '') == 'legacy':
+            data = asarray(a)
+        else:
+            data = ravel(asarray(a))
```
Can we just ditch the `ravel` here altogether? I don't think any of the builtin formatters actually care.

ahaldane replied:
At least `IntegerFormat` does, since it used `np.maximum.reduce` to find the largest int:
```python
>>> x = np.array([[1, 2], [3, 4]])
>>> np.maximum.reduce(x)
array([3, 4])
>>> np.maximum.reduce(np.ravel(x))
4
```

eric-wieser replied:
Changing that to `np.max` would fix that

ahaldane commented:
I split off the recursionguard fixes to #9688

ahaldane committed:
MAINT: arrayprint.py formatters no longer require ravel'd data

ahaldane commented:
OK, replaced all uses of `np.minimum.reduce` by `np.min`, and same for max, and removed ravel in a few places.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -305,24 +298,22 @@ def indirect(x):
         for field_name in dtype_.names:
-            field_values = data[field_name]
-            format_function = _get_format_function(
-                    ravel(field_values), precision, suppress_small, formatter)
```
This change affects formatting void scalars containing floats, right? (in a good way)

eric-wieser replied:
I think `np.array((1.0,), np.dtype([('a', float)]))` might be the test, although I'm not sure what that did on your previous commit. On 1.13, that returns `( 1.0,)` with the space

ahaldane replied:
I don't quite follow.. in these particular lines (316-318) the only functional change is to remove a call to `ravel`. 

In 1.13 it is true that void scalars would never get here, however in 1.13 these 3 lines were copied in the `()` special case in `array2string`, so 1.13 ended up doing the same thing.

It's also true that spaces within void scalars and void arrays are gone now. I will add a test for that.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -607,15 +603,21 @@ def _formatArray(a, format_function, rank, max_line_len,
         try:
-            self.fillFormat(data)
-        except (TypeError, NotImplementedError):
+            self.fillFormat(ravel(data))
```
Is this `ravel` needed?

ahaldane commented:
Removed the last uses of ravel, added a test for structured types.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -607,15 +602,21 @@ def _formatArray(a, format_function, rank, max_line_len,
         try:
             self.fillFormat(data)
-        except (TypeError, NotImplementedError):
+        except (NotImplementedError):
```
Parens here are weird

ahaldane replied:
Will do once we have decided whether there are other changes to make (your other comment).

By the way, the reason we can remove the TypeError is because of the addition of all the lambdas in #8963.

ahaldane replied:
Oh actually, on second look, I don't think we need the `NotImplementedError` check either, so this whole try/catch can be removed.

ahaldane replied:
I'm also going to move all of the imports of `from . import numeric as _nc` that appear at the start of 3 or 4 functions, and just make that a global import. There doesn't seem to be any recursive import problem.

eric-wieser replied:
You might want to check if we have a more standard name for that than `_nc` - I think I've seen `_nx` in a bunch of places.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -734,29 +734,51 @@ def __call__(self, x):
+        # add an extra space so " True" and "False" have the same length and
+        # array elements align nicely when printed, but only for arrays with
+        # more than one element (0d and nd)
+        self.truestr = ' True' if data.size > 1 else 'True'
```
I don't remember the rationale here - is there any reason not to change this to `.shape == ()`, now that we don't `ravel`?

That is, is it a deliberate choice to change to `array([True])` instead of `array([ True])` for anything but the scalar case?

I do vaguely recall deciding _not_ to use `not data.all()` as a condition here, and I think using `()` here would be consistent with that choice

ahaldane replied:
It is a deliberate choice, but debate is welcome.

It is so that the 1d array constructed as `np.array([True])` prints as `np.array([True], dtype=bool)` (without a space) and not `np.array([ True], dtype=bool)`. If we changed the test to `data.shape == ()` then this array would be printed with a space, which I thought was slightly ugly without any other benefit.

Perhaps, as you argue, the benefit would be consistency with longer boolean ndarrays, which print with a space, eg `array([ True,  True], dtype=bool)`. The spaces are added so that if a user prints out multiple boolean arrays on different lines of output, then the elements are aligned with each other across lines, for easy reading. But I didn't think this was useful if we are only printing size-1 arrays, since there is no reading problem.

To me the slight ugliness of the space outweighs the lack of consistency, but say the word and I can change it.

eric-wieser replied:
I slightly prefer the consistency here - it's also one fewer case that users downstream would need to change their doctests for.

I'd say either wait for a third opinion to break the tie (@mhvk?), or change it so I can merge sooner. It'd be trivial to change this in a later PR anyway, as long as it still hits 1.14

mhvk replied:
Hmm, tie-breaker, me? I think my preferred alternative was to omit the space for any case in which the array was all True, in analogy with the case where all numbers are positive...

eric-wieser reviewed (commented):
So close! Very happy with the implementation now - thanks for removing all the `ravel`s.

Just a small nit and a design question for boolean arrays left:

ahaldane committed:
MAINT: fixups to legacy spacing in FloatFormat

ahaldane commented:
Updated, all updates are in the last commit.

Summary:
 * made size-1 nd bool arrays use a space before True, giving `array([ True], dtype=bool)`
 * in `legacy` sign mode, kept extra space in the exp float format, plus tests for this
 * made the _nx import global
 * removed NotImplemented try/catch in FloatFormat

ahaldane commented:
OK, turns out there was a circular import problem in python2. 

Fixed by moving the `array_str` and `array_repr` code from `numeric.py` to `arrayprint.py`, which seems like where it should be anyway.

ahaldane commented:
Moved the circular import changes to another PR #9725 , to avoid blowing this one up.

It's probably cleaner to merge that one first, in case we want to revert this one.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -610,30 +605,38 @@ def _formatArray(a, format_function, rank, max_line_len,
-                min_val = minimum.reduce(non_zero)
+                max_val = np.max(abs_non_zero)
+                min_val = np.min(abs_non_zero)
+                min_val_sgn = np.min(non_zero)
```
Seems a bit odd that we're importing `np` directly for these, yet importing other stuff indirectly. Not something you need to fix in this PR though

ahaldane replied:
Yes, looks like we can replace all the indirect imports can be replaced by the single `np` import - I just tried it, seems fine.

I won't do that in this PR though.

charris replied:
I've never understood quite how, or why, that worked. But it generally does.

ahaldane replied:
I think it works because we never use or call any numpy functions or variables in this file at import time, only at runtime. 

At the moment `arrayprint` is imported from the main`__init__` file, the `numpy` module has been created but not populated, so we are allowed to import numpy, but not allowed refer to any of its functions or variables. However, we can use it inside of function definitions since by the time they are called the numpy module will have been populated.

(Actually, there I see exceptions: At module import time we call `multiarray.set_string_function` and we refer to `int_`, `float_`, `complex_`, `intc` and `longlong`. Therefore these will still need to be directly imported).

ahaldane replied:
Hmm, but it might be safer to leave these all the way they are: What if another module calls arrayprint functions at import time? Then we would have a problem. Ideally no other modules would, though.

(And this isn't theoretical: We recently had a problem because `getlimits.py`was calling `array2string` at module import time, though we've now fixed it so it no longer does (#9113))

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -733,29 +732,46 @@ def __call__(self, x):
+    def __init__(self, data, **kwargs):
+        # add an extra space so " True" and "False" have the same length and
+        # array elements align nicely when printed, but only for arrays with
+        # more than one element (0d and nd)
```
Comment no longer matches code

ahaldane commented:
Fixed the comment, plus the two doctrings describing the `sign` argument.

eric-wieser reviewed (approved):
LGTM! Thanks for the iteration.

Let's leave this a day or two to collect any final comments

eric-wieser commented:
@mhvk: good to go from your perspective?

mhvk commented:
Looked over it once more and, yes, good to go! Though we should ensure that `LongFloat` gets done before 1.14 as well...

ahaldane commented:
I just tweaked the release note a little. I changed the title about `Unneeded whitespace in float and bool array printing removed` to `Unneeded whitespace in float array printing removed` and edited the note below.

ahaldane commented:
Otherwise, I checked over it again, I don't have more changes.

eric-wieser commented:
> I intend to put this in just to see who screams ;) 

@charris, I think it's screaming time

ahaldane commented:
Rebased for conflicts with #9688

charris commented:
In it goes. Good job all.

charris commented:
There look to be four remaining PR relevant to this. It would be helpful if folks could keep Allan's list current and close any PRs that are no longer relevant.
